pycbc/inference/evidence.py

```diff
@@ -0,0 +1,465 @@
+    formally has infinite variance, very error prone when log
+    likelihood values are small relative to the mean log likelihood.
+
+    Not recommended for general use.
+
+    Parameters
+    ----------
+    log_likelihood : 1d array of floats
+                The log likelihood of the data sampled from the
+                posterior distribution.
+
+    Returns
+    ----------
+    log_evidence   : float
+                Estimation of the log of the evidence.
+    """
+    num_samples = len(log_likelihood)
+    logl_max = numpy.max(-1.0*log_likelihood)
+
+    log_evidence = 0.
+    for i in range(len(log_likelihood)):
+        log_evidence += numpy.exp(-1.0*log_likelihood[i] + logl_max)
+
+    log_evidence = -1.0*numpy.log(log_evidence)
+    log_evidence += logl_max
+    log_evidence += numpy.log(num_samples)
+
+    return log_evidence
+
+def trapezoid_TI(log_likelihood, betas):
+    """
+    Returns the evidence of the data via thermodynamic integration
+    with the trapezoid rule. It also returns an estimate of the MCMC
+    standard deviation of the evidence. 
+
+    Parameters
+    ----------
+
+    log_likelihood : 3d array of shape (betas, walker, iteration)
+                The log likelihood for each temperature separated
+                by temperature, walker, and iteration.
+
+    betas          : 1d array
+                The inverse temperatures used in the MCMC.
+
+    Returns
+    ----------
+    log_evidence   : float
+                Estimation of the log of the evidence.
+
+    mcmc_std       : float
+                The standard deviation of the evidence estimate from
+                Monte-Carlo spread.
+    """
+    # ascending order sort
+    order = numpy.argsort(betas)
+    betas = betas[order]
+    log_likelihood = log_likelihood[order]
+
+    # Assume log likelihood is given in shape of beta,
+    # walker, iteration.
+    log_likelihood = numpy.reshape(log_likelihood, (len(betas),
+                                                    len(log_likelihood[0].flatten())))
+
+    log_evidence = 0
+    for i in range(len(betas) - 1):
+        delta_beta = betas[i+1] - betas[i]
+        pre_fac_mean = (1. / 2.) * delta_beta
+        log_evidence +=  pre_fac_mean * (numpy.average(log_likelihood[i+1]) + numpy.average(log_likelihood[i]))
+
+    # Estimate the MCMC variation
+    TI_vec = numpy.zeros(len(log_likelihood[0]))
+
+    # Estimating average thermodynamic integration
+    # of the evidence along with a method for estimating the
+    # standard deviation.
+    # See https://www.sciencedirect.com/science/article/pii/S0022249617302651
+    # Eqs  (36) and (37)
+    for i in range(len(log_likelihood[0])):
+        log_likelihood_samps = [log_likelihood[x][i] for x in range(len(betas))]
+        TI_vec[i] = numpy.trapz(log_likelihood_samps, betas)
+
+    mcmc_std = numpy.std(TI_vec) / numpy.sqrt(float(len(log_likelihood[0])))
+
+    return log_evidence, mcmc_std
+
+def improved_trapezoid_TI(log_likelihood, betas):
```

Comment:
Anything more descriptive than "improved"?

Reply:
Yeah, I'll fix that.

higher_order_trapezoid maybe? https://en.wikipedia.org/wiki/Trapezoidal_rule#Error_analysis

It's adding in that constant term.